connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala

```diff
@@ -312,14 +368,11 @@ private[sql] class AvroSerializer(
    */
   private def resolveAvroType(avroType: Schema): (Boolean, Schema) = {
     if (avroType.getType == Type.UNION) {
-      val fields = avroType.getTypes.asScala
-      val actualType = fields.filter(_.getType != Type.NULL)
-      if (fields.length != 2 || actualType.length != 1) {
-        throw new UnsupportedAvroTypeException(
-          s"Unsupported Avro UNION type $avroType: Only UNION of a null type and a non-null " +
-            "type is supported")
+      nonNullUnionBranches(avroType) match {
```

Comment:
This logic has a couple of issues determining nullability for a union. Checking if the union contains a null is needed to handle these cases.

1. Case `["null", "int", "long"]` will hit line 374 and return `nullable=false` when it should return true.
2. Case `["int"]` will hit line 373 and return `nullable=true` when it should return false.

Maybe something like this?

```scala
val containsNull = avroType.getTypes.asScala.exists(_.getType == Schema.Type.NULL)

nonNullUnionBranches(avroType) match {
    case Seq() => (true, Schema.create(Type.NULL))
    case Seq(singleType) => (containsNull, singleType)
    case _ => (containsNull, avroType)
}
```